fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnectorScanRangeSource.java

```diff
@@ -0,0 +1,409 @@
+                        .setPriority(dataCacheRule.get().getPriority()).build());
+            }
+        } else {
+            // evaluate partition predicates
+            for (PartitionKey partitionKey : partitionKeys) {
+                // key is ColumnName, value is Expr(Literal)
+                Map<String, Expr> mapping = new HashMap<>(partitionColumnNames.size());
+                Preconditions.checkArgument(partitionColumnNames.size() == partitionKey.getKeys().size(),
+                        "PartitionColumnName size must equal with PartitionKey keys' size.");
+                for (int i = 0; i < partitionKey.getKeys().size(); i++) {
+                    mapping.put(partitionColumnNames.get(i), partitionKey.getKeys().get(i));
+                }
+                // Must clone expr first, avoid change original expr
+                Expr clonedExpr = predicates.clone();
+                Expr rewritedExpr = DataCacheExprRewriter.rewrite(clonedExpr, mapping);
+                ScalarOperator op = SqlToScalarOperatorTranslator.translate(rewritedExpr);
+                ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
+                op = scalarRewriter.rewrite(op, ScalarOperatorRewriter.DEFAULT_REWRITE_RULES);
+                if (op.isConstantTrue()) {
+                    // matched partition predicates
+                    dataCacheOptions.add(DataCacheOptions.DataCacheOptionsBuilder.builder()
+                            .setPriority(dataCacheRule.get().getPriority()).build());
+                } else {
+                    // not matched, add null DataCacheOption
+                    dataCacheOptions.add(null);
+                    if (!op.isConstantRef()) {
+                        LOG.warn(String.format("ConstFolding failed for expr: %s, rewrite scalarOperator is %s",
+                                rewritedExpr.toMySql(), op.debugString()));
+                    }
+                }
+            }
+        }
+        return Optional.of(dataCacheOptions);
+    }
+
+    class ScanRangeIterator implements Iterator<TScanRangeLocations> {
+        RemoteFileInfo remoteFileInfo;
+        List<RemoteFileDesc> files;
+        int fileIndex = 0;
+        List<TScanRangeLocations> buffer = new ArrayList<>();
+
+        public ScanRangeIterator(RemoteFileInfo remoteFileInfo) {
+            this.remoteFileInfo = remoteFileInfo;
+            this.files = remoteFileInfo.getFiles();
+        }
+
+        @Override
+        public boolean hasNext() {
+            tryPopulateBuffer();
+            return fileIndex < files.size() || !buffer.isEmpty();
+        }
+
+        @Override
+        public TScanRangeLocations next() {
+            return buffer.remove(buffer.size() - 1);
+        }
+
+        private void tryPopulateBuffer() {
+            while (buffer.isEmpty() && fileIndex < files.size()) {
+                splitFile(files.get(fileIndex));
```

Comment:
buffer is not empty after splitFile, so just populate one file?

Reply:
yes.  try to split a singe files into scan ranges everytime.